### DISCUSSION TITLE
Inquiry about the general workflow of how the cross-linking of identifiers and chemical information assigned in humanGEM

### DISCUSSION THREAD

**Discussion by gmhhope** [SysBioChalmers/Human-GEM#370] [Q&A] [answered]
Dear Human-GEM developers,

I am very happy to see the constant updates of Human-GEM model and thank you for making this model available! I am currently compiling the model for our usage. What I try to understand is how you assign all the crosslinking identifiers and accurately assign the chemical information. 

- If I understand correctly and based on the paper, most of the cross-linking of compound identifiers comes from MetNetX. Is that correct? 
- I do see some orphan identifiers coming from VMH (e.g., 'vmhmetabolite': 'CE7087'). Do you know where can I find the full tables of these `vmh` metabolites and their formula and charge information?
- How the formula and charge were assigned? Do they mostly come from MetNetX `chemProp.tsv`? 

I am sorry that I didn't follow your format. Hope you can help!

Thanks,
Minghao Gong

**Comment by JonathanRob** [comment, marked as answer]
Hi @gmhhope, thank you for your comments, we're glad you find the resource useful!

The mapping and assignment of identifiers has not been so systematic, given the incompleteness and sometimes questionable reliability of the many sources (previous GEMs, databases, etc.). This means that there is a lot of manual work involved which is neither efficient nor sustainable long-term, so we're still in the process of trying to automate as much of it as possible.

> If I understand correctly and based on the paper, most of the cross-linking of compound identifiers comes from MetNetX. Is that correct?

I believe most of the identifier information currently in Human-GEM (and its associated annotation files) actually comes from its predecessor models (HMR, Recon, and iHsa), during a very careful and involved merging process. We did attempt to map as many identifiers to MetaNetX as it was one of the most exhaustive and comprehensive resources we found, and this was in turn used to retrieve some additional identifiers and chemical information, and resolve some conflicts.

The plan moving forward is to use the **ChEBI ID** as the "true source" when possible, and eventually use that to retrieve cross-linked identifiers and chemical properties. But of course not all of the model metabolites yet (or will ever) exist in the ChEBI database, so **MetaNetX** mapping will also continue to be a high priority.

> I do see some orphan identifiers coming from VMH (e.g., 'vmhmetabolite': 'CE7087'). Do you know where can I find the full tables of these `vmh` metabolites and their formula and charge information?

Great question - I'm actually not sure how to obtain some sort of flatfile (e.g., csv) that includes the VMH metabolite identifiers and associations/chemical information from the VMH website. Maybe @mihai-sysbio knows?

> How the formula and charge were assigned? Do they mostly come from MetNetX `chemProp.tsv`?

Most originate from the preceding models. We also did an extensive manual curation of metabolite formulas and charges after generating Human1 to reconcile mass/charge imbalances and address problematic cases such as polymerization and highly combinatorial processes (e.g., lipid metabolism). This unfortunately means that some of the formulas may not agree with some databases, but usually those cases are for quite large/complex metabolites. The assumed pH/protonation state is also a major factor contributing to differences in formula/charge among different sources, further complicated a more automated handling of this issue.

**Reply by mihai-sysbio** [reply]
Without being entirely sure what you're after @gmhhope, I had a quick go at it https://colab.research.google.com/drive/1XL7Xqcw8TVgiLdwZS_uaXL9Dl4GxSXr6

**Reply by gmhhope** [reply]
OMG, thanks Jonathan and Mihai for your involved explanations!

I am sorry that I lost track of this during last week. I am going to look deeper and let you know if I resolve my problems.

Thanks,
Minghao Gong

**Reply by gmhhope** [reply]
Hi Jonathan and Mihai,

Thanks again for the answers. Actually, I am trying to build our own metabolite databases that link metabolic models (e.g., humanGEM) with **metabolomic-centric applications**. That is why I am looking after identifier matching and formula deduction. In Mass spectrometry-based application, the neutral monoisotopic mass/formula is important. But in GEM metabolic model, the charged formula is essential. Thus, there are issues to translate between them. Even though I may guess right a neutral formula from a charged one by some deduction, I still cannot get more chemical information if the entries from GEMs don't have identifiers linking to the chemical databases (e.g., CHEBI). Thus, I am very curious how you guys link the external identifiers between your model and the public chemical databases like CHEBI. 

Now it sounds like most of them are coming from the previous models (HMR, Recon, and iHsa)? And then you may try to enrich the identifiers by using MetNetX? I realized that MetNetX also did not include Recon3D/VMH identifiers yet. That may explain why your humanGEM model has many orphan identifiers of VMH. This is quite problematic as we are also trying to parse some microbial models from VMH but then found more than half of them don't have external identifiers. If you guys happen to find the flatfile or json files that document all their metabolite entries, please also let me know!

BTW, I am also looking into CHEBI and try extracting the IDs with neutral formula, inchikey and monoisotopic mass. If you know, do they provide the monoisotopic mass? I check in there (e.g., `chemical_data.tsv` from https://ftp.ebi.ac.uk/pub/databases/chebi/Flat_file_tab_delimited/ but can find formula but not monoisotopic mass there. And I don't know those formula are charged or not. If they are charged, it will be great for GEM but become problematic for MS-based applications.

Thanks again for your meticulous responses to my questions. I look forward to your angles to think about this question. 

Best,
Minghao Gong

**Reply by mihai-sysbio** [reply]
A metabolite database is a very interesting avenue to explore @gmhhope. When brainstorming potential future uses of [Metabolic Atlas](https://metabolicatlas.org), this has been one of the ideas that came up. My (unsolicited) suggestion when building any database that uses models is to, at least in the beginning, focus on models which can improve with time, i.e. they have (often) releases; this makes the resulting database better in time, without necessarily putting more effort into it, if the model incorporation is automatic.

The MNX mapping has been essential to enrich the annotation, and I think @JonathanRob and @haowang-bioinfo can fill in how this was done before that (it's probably detailed in some of the old PRs).

 @gmhhope I might have found a solution regarding fetching the metabolite details. From the examples I looked at, for every Recon3D-sourced metabolite in Human-GEM there is an associated HMDB identifier that contains the expected metabolite details, at least on their website (see the [updated notebook](https://colab.research.google.com/drive/1XL7Xqcw8TVgiLdwZS_uaXL9Dl4GxSXr6)).

**Reply by haowang-bioinfo** [reply]
@gmhhope thanks for bringing up these important questions both for users and ourselves. As @JonathanRob explained, we paid tremendous efforts in cross-linking those identifiers, also welcome contributions and bug report from the community.

Among the early releases ([v0.5.1](https://github.com/SysBioChalmers/Human-GEM/releases/tag/v0.5.1) and before), MNX-based identifier mapping steps were implemented and exemplified as in this [script](https://github.com/SysBioChalmers/Human-GEM/blob/main/.deprecated/code/modelCuration/MetAssociation/curateHMR2Mets.m). All code previously used for mapping [metabolites](https://github.com/SysBioChalmers/Human-GEM/tree/main/.deprecated/code/modelCuration/MetAssociation) and [reactions](https://github.com/SysBioChalmers/Human-GEM/tree/main/.deprecated/code/modelCuration/RxnAssociation) has been reserved as log for tracking changes. 

In the most recent releases, we are moving toward relying on the [ChEBI](https://www.ebi.ac.uk/chebi/) database as trustable metabolite source with the kind support from ChEBI curator @ANiknejad.

We hope Human-GEM can server the community as a robust and reliable resource for studying human metabolism and biomedical research.

**Reply by gmhhope** [reply]
Thanks very much! @mihai-sysbio @haowang-bioinfo,

All your suggestions and comments are very helpful information! 

@mihai-sysbio, `From the examples I looked at, for every Recon3D-sourced metabolite in Human-GEM there is an associated HMDB identifier that contains the expected metabolite details, at least on their website`. I have looked into your notebook and I appreciate a lot for that! I found out that in your script, because of the `try` clause, only those 478 entries with valid `HMDB` field are retained. Otherwise, it will not retain. And when I remove that, definitely a large amount of those don't have HMDB identifiers. But thanks for trying that out! I really appreciate it!

Also, you refer to a website that may be relevant, could you share me the link?

I am very happy to be involved in such an active community. Hope I can contribute soon!

Thanks a lot,
Minghao Gong

**Reply by gmhhope** [reply]
Also, can I assume that the the list of metabolites and their identifiers in https://github.com/SysBioChalmers/Human-GEM/blob/main/model/Human-GEM.xml will be always synchronized with https://github.com/SysBioChalmers/Human-GEM/blob/main/model/metabolites.tsv

Thanks,
MInghao Gong

**Reply by haowang-bioinfo** [reply]
> Also, can I assume that the the list of metabolites and their identifiers in https://github.com/SysBioChalmers/Human-GEM/blob/main/model/Human-GEM.xml will be always synchronized with https://github.com/SysBioChalmers/Human-GEM/blob/main/model/metabolites.tsv

yes, indeed

**Reply by smoretti** [reply]
@gmhhope some VMH compound information can be found here: https://github.com/opencobra/ctf but not everything I guess

ChEBI does provides the Monoisotopic Mass (e.g. https://www.ebi.ac.uk/chebi/searchId.do?chebiId=169172). It should be available in the ChEBI OWL or OBO download files: https://ftp.ebi.ac.uk/pub/databases/chebi/ontology/

**Reply by gmhhope** [reply]
Thanks @smoretti !

The information is indeed uncharted territory for me. Thanks for providing such valuable information!

Best,
Minghao
